druid-shell/src/platform/x11/window.rs

```diff
@@ -401,30 +355,37 @@ impl Window {
     }
 
     fn render(&self, invalid_rect: Rect) -> Result<(), Error> {
-        // TODO(x11/errors): this function should return a an error
-        // instead of panicking or logging if the error isn't recoverable.
-
         // Figure out the window's current size
         let geometry_cookie = xcb::get_geometry(self.app.connection(), self.id);
-        let reply = geometry_cookie.get_reply().unwrap();
+        let reply = geometry_cookie.get_reply()?;
         let size = Size::new(reply.width() as f64, reply.height() as f64);
         self.set_size(size)?;
 
         let mut anim = false;
-        if let Ok(mut cairo_ctx) = self.cairo_context.try_borrow_mut() {
+        {
+            let mut cairo_ctx = borrow_mut!(self.cairo_context)?;
             let mut piet_ctx = Piet::new(&mut cairo_ctx);
             piet_ctx.clip(invalid_rect);
-            if let Ok(mut handler) = self.handler.try_borrow_mut() {
-                anim = handler.paint(&mut piet_ctx, invalid_rect);
-            } else {
-                log::error!("Window::render - handler already borrowed");
-            }
-            if let Err(e) = piet_ctx.finish() {
-                log::error!("Window::render - piet finish failed: {}", e);
-            }
+
+            // We need to be careful with earlier returns here, because piet_ctx
+            // can panic if it isn't finish()ed. Also, we want to reset cairo's clip
+            // even on error.
+            let err;
+            match borrow_mut!(self.handler) {
+                Ok(mut handler) => {
+                    anim = handler.paint(&mut piet_ctx, invalid_rect);
+                    err = piet_ctx
+                        .finish()
+                        .map_err(|e| anyhow!("Window::render - piet finish failed: {}", e));
+                }
+                Err(e) => {
+                    err = Err(e);
+                    let _ = piet_ctx.finish();
+                }
+            };
```

Comment:
The second `finish` call could fail as well, right? If we can add it to `err`, than that would seem useful, otherwise we could just keep `log`ing this one.

Reply:
The only way I can find to return a double error is `context`, which isn't really appropriate IMO because it's for when one error is caused by another. I added a log, though.